zephyr/test/vmh.c

```diff
@@ -17,8 +18,258 @@
 
 LOG_MODULE_DECLARE(sof_boot_test, CONFIG_SOF_LOG_LEVEL);
 
+/* Test creating and freeing a virtual memory heap */
+static void test_vmh_init_and_free_heap(int memory_region_attribute,
+					struct vmh_heap_config *config,
+					int core_id,
+					bool allocating_continuously,
+					bool expect_success)
+{
+	struct vmh_heap *heap = vmh_init_heap(config, memory_region_attribute,
+		core_id, allocating_continuously);
+	if (expect_success) {
+		zassert_not_null(heap,
+		"Heap initialization expected to succeed but failed");
+		}
+	else
+		zassert_is_null(heap, "Heap initialization expected to fail but succeeded");
+
+	if (heap) {
+		int ret = vmh_free_heap(heap);
+
+		zassert_equal(ret, 0, "Failed to free heap");
+	}
+}
+
+/* Test for vmh_alloc and vmh_free */
+static void test_vmh_alloc_free_no_check(struct vmh_heap *heap,
+					uint32_t alloc_size,
+					bool expect_success)
+{
+	void *ptr = vmh_alloc(heap, alloc_size);
+
+	if (expect_success)
+		zassert_not_null(ptr, "Allocation expected to succeed but failed");
+	else
+		zassert_is_null(ptr, "Allocation expected to fail but succeeded");
+
+	if (ptr) {
+		int ret = vmh_free(heap, ptr);
+
+		zassert_equal(ret, 0, "Failed to free allocated memory");
+	}
+}
+
+/* Test function for vmh_alloc and vmh_free with memory read/write */
+static void test_vmh_alloc_free_check(struct vmh_heap *heap,
+					uint32_t alloc_size,
+					bool expect_success)
+{
+	void *ptr = vmh_alloc(heap, alloc_size);
+
+	if (expect_success)
+		zassert_not_null(ptr, "Allocation expected to succeed but failed");
+	else {
+		zassert_is_null(ptr, "Allocation expected to fail but succeeded");
+		return;
+	}
```

Comment:
no. I don't think you need a `return` after an assert, asserts **are** checks themselves, so all this could be a single line like
```
__ASSERT(expect_success == !!ptr, "...");
```